R/stat-.r

```diff
@@ -97,10 +97,18 @@ Stat <- ggproto("Stat",
     args <- c(list(data = quote(data), scales = quote(scales)), params)
     dapply(data, "PANEL", function(data) {
       scales <- layout$get_scales(data$PANEL[1])
-      try_fetch(do.call(self$compute_panel, args), error = function(cnd) {
-        cli::cli_warn("Computation failed in {.fn {snake_class(self)}}", parent = cnd)
-        new_data_frame()
-      })
+      try_fetch(do.call(self$compute_panel, args),
```

Comment:
Unless I'm missing something, I think this should be more like:

```R
try_fetch(do.call(self$compute_panel, args),
  rlib_error_package_not_found = function(cnd) {
    cli::cli_abort("Aborted computation in {.fn {snake_class(self)}}", parent = cnd)
  },
  error = function(cnd) {
    cli::cli_warn("Computation failed in {.fn {snake_class(self)}}", parent = cnd)
    new_data_frame()
  }
)
```

Reply:
I thought so at first, but the rethrown error gets caught in the `error` case, so I had to do it all in `error` case. Are there any better way?

``` r
ignore_all_but_custom_error <- function(expr) {
  result <- rlang::try_fetch(
    expr,
    custom_error = function(cnd) {
      rlang::inform("custom_error detected")
      rlang::abort("rethrowing the error", parent = cnd)
    },
    error = function(cnd) {
      rlang::inform("ignoring the error")
    },
  )
  
  invisible(result)
}

# works as expected
ignore_all_but_custom_error(stop("foo"))
#> ignoring the error

# hmm...
ignore_all_but_custom_error(rlang::abort("foo", class = "custom_error"))
#> custom_error detected
#> ignoring the error
```

<sup>Created on 2022-05-20 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
Hmmm, I suspect this is because of the way that `tryCatch()` adds the calling handlers behind the scenes. Hopefully @lionel- can suggest a fix.

Reply:
Now I'm wondering if `check_installed()` should signal non-error conditions. This way it'd work in `tryCatch()`.

There are two steps currently:

1. Signal a classed condition (that currently inherits from `"error"`) with a restart.
2. Call `invokeRestart("abort")` if user selected no, bypassing everything on the stack including error handlers.

The second step is `tryCatch()`-proof but not the first step. If it were, I think no complications would be needed.

Reply:
But doesn't the same problem arise with any classed condition? i.e. you can't use `tryCatch()` to handle a classed error _and_ all other errors.

Reply:
> But doesn't the same problem arise with any classed condition?

If we use a classed condition that doesn't inherit from `"error"` to communicate, `"error"` handlers won't be called when we signal. This way we avoid interactions with catch-all-errors handlers.

>  i.e. you can't use tryCatch() to handle a classed error and all other errors.

Sorry I'm not sure I understand what you mean.

Reply:
I think we have to keep the error class as errors, otherwise they wouldn't get caught by a generic `error` argument? 

Isn't the whole point of the design of `tryCatch()` to allow you to deal with error subclasses individually or as a whole?

Reply:
I was wondering whether check-installed really follows the error model or instead is more like an interrupt.

- Like interrupts, the current computation is aborted after user interaction, when the user selects "No". By contrast errors are triggered by the program rather than the user.

- Like interrupts, this is accomplished using `invokeRestart("abort")`, at which point nothing can prevent the jump (except an intercepting `"abort"` restart).

Originally I restarted to `"abort"` instead of throwing an error to improve the UI. When the user selects no, they are taken back to a prompt instead of seeing an error.

Also note that currently `tryCatch(error = )` is causing `check_installed()` to exit too soon, I think it's a bug. We added a restart mechanism in r-lib/rlang#1199 in which we signal the error condition instead of a bare condition. I think this is not correct since condition communication should always be done with bare conditions (or more precisely classed conditions that don't inherit from error, warning, or message) to avoid catch-all handlers.

We can fix the restart signal protocol but then we still need to think about what UI do we want to implement:

1. Should `check_installed()` prompt the user at all within `tryCatch(error = )`?

2. If we do prompt the user, and they select "No", should they consistently be taken back to the prompt even when within `tryCatch()`? In that case, this means we implement the interrupt model in interactive sessions. Otherwise, we should resignal the condition, this time inheriting from `"error"`.

Regarding (1), I think we should prompt the user in interactive sessions, since that is the documented purpose of `check_installed()`. Though I'm a bit concerned about users setting up a loop e.g. with `safely()` to run a long computation and then come back to their computer the next day to find the `check_installed()` prompt. Maybe we should detect that we are called within `source()` and in that case behave as if the session was non-interactive?

Regarding (2), I think the interrupt model makes the most sense to me. I'd be a bit surprised to select "No" and then see the computation continue.

Reply:
Ok, that makes sense — that way the initial call is an interrupt and it only becomes an error if you select "No". I'm not too concerned with your `safely()` scenario, because I think it will be relatively unlikely that a computation half-way through the loop will require package installation, and generally you will have prototyped at least once before doing the loop. 

I agree that selecting "No" should take you to the top level, as otherwise in the code below you'd need to select "No" 10 times in order to exit the `map()`:

```R
f <- function(i) {
  # make a plot that uses hexbin
  check_installed("hexbin")
}
purrr::map(1:10, safely(f))
```

Reply:
> I agree that selecting "No" should take you to the top level,

hmm but then we agree that `check_installed()` is either a no-op (after a restart if user selects yes) or an interrupt (if no is selected), never an error, right?

Reply:
Great point regarding repeated prompts in loops by the way. This clearly shows why `check_installed()` should implement the interrupt model rather than the error model.